File: src/Assets/Modules/Trunk/Agent/Samurai/SamuraiFocus.cs
//using GameLib.DI;
//using QS.Chara;
//using QS.Common.Util;
//using UnityEngine;

//namespace QS.Agent
//{
//    public class SamuraiFocus : AgentStateBase
//    {

//        public override AgentState State => AgentState.Focus;

//        [Injected]
//        readonly IAimLockInstr aimLockInstr;
//        public SamuraiFocus(SamuraiAgent agent) : base(agent)
//        {
//            aimLockInstr = AgentGlobal.Instance.GetInstance<IAimLockInstr>();
//        }
//        public override void Enter()
//        {
//            base.Enter();
//            aimLockInstr.Target = agent.Sensor.Enemy;
//            agent.Chara.Execute(aimLockInstr);
//        }
//        // 孙子状态用ifelse 放Lua实现
//        public override void Process()
//        {
//            var randomVal = Random.Range(0, 100);
//            if (!agent.Sensor.EnemyFound) return;

//            if (!TransformUtil.IsInSectorArea(agent.transform, agent.Sensor.Enemy, Common.RelativeDir.Forward, 100, 2.5f))
//            {
//                agent.Actions.TryAdd(new ApprochAction(agent, agent.Sensor.Enemy));
//            }
//            else if (agent.Actions.ActiveAction is KatanaLightAttackAction)
//            {
//                if (randomVal < 95)
//                {
//                    agent.Actions.TryAdd(new RecedeAction(agent, agent.Sensor.Enemy));
//                }
//                else if (Time.time - agent.Actions.ActiveAction.ExecutedTime < 0.75)
//                {
//                    agent.Actions.TryAdd(new KatanaLightAttackAction(agent, agent.Sensor.Enemy));
//                }
//            }
//            else if (randomVal < 80)
//            {
//                agent.Actions.Clear();
//                agent.Actions.TryAdd(new KatanaLightAttackAction(agent, agent.Sensor.Enemy));
//            }



//        }
//    }
//}

File: src/Assets/LuaScript/GameAI/Samurai/Samurai_Focus.lua
--[[--
好了，是时候来设计AI脚本的API了。AI脚本用于定义单独的一个子状态
Lua 命名规范
函数 小驼峰
变量 蛇形命名
常量 全大写，下划线分割
--]]--

local actions = require 'actions'
local instrs = require 'instrs'

local aim_lock_instr = instrs.create(instrs.CHARA_AIM_LOCK)

function enter()
    aim_lock_instr.Target = self.Sensor.Enemy
    DebugLog(aim_lock_instr)
    self.Chara:Execute(aim_lock_instr)
end

function process()
    math.randomseed(os.time())
    local randomval = math.random(100)

    if not self.Sensor.EnemyFound then
        return
    end

    if not TransformUtil.IsInSectorArea(self.transform, self.Sensor.Enemy, RelativeDir.Forward, 100, 3) then
        self.Actions:TryAdd(actions.create(actions.COMMON_APPROCH, self, self.Sensor.Enemy))
    elseif randomval < 10 then
        self.Actions:Clear()
        local dir = -1
        if math.random(100) < 50 then
            dir = -1
        else
            dir = 1
        end
        self.Actions:TryAdd(actions.create(actions.COMMON_DODGE, self, Vector2(dir, 0), 1 ,3))
    elseif randomval < 70 then
        self.Actions:TryAdd(actions.create(actions.SK_LIGHT_ATTACK, self, self.Sensor.Enemy))
    elseif randomval > 90 then
        self.Actions:Clear()
        self.Actions:TryAdd(actions.create(actions.COMMON_RECEDE, self, self.Sensor.Enemy))
    end

end